core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -79,13 +75,71 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
+
+    }
+    protected RoutingAllocation.Result buildChangedResult(MetaData metaData, RoutingNodes routingNodes, RoutingExplanations explanations) {
+        final RoutingTable routingTable = new RoutingTable.Builder().updateNodes(routingNodes)
+                .version(routingNodes.getRoutingTable().version() + 1).build();
+        MetaData newMetaData = updateMetaDataWithRoutingTable(metaData,routingTable);
+        return new RoutingAllocation.Result(true, routingTable.validateRaiseException(newMetaData), newMetaData, explanations);
+    }
+
+    /**
+     * Updates the current {@link MetaData} based on the newly created {@link RoutingTable}.
+     *
+     * @param currentMetaData {@link MetaData} object from before the routing table was changed.
+     * @param newRoutingTable new {@link RoutingTable} created by the allocation change
+     * @return adpated {@link MetaData}, potentially the original one if no change was needed.
+     */
+    static MetaData updateMetaDataWithRoutingTable(MetaData currentMetaData, RoutingTable newRoutingTable) {
+        // make sure index meta data and routing tables are in sync w.r.t primaryTerm
+        MetaData.Builder metaDataBuilder = null;
+        for (IndexRoutingTable indexRoutingTable : newRoutingTable) {
+            final IndexMetaData indexMetaData = currentMetaData.index(indexRoutingTable.getIndex());
+            if (indexMetaData == null) {
+                throw new IllegalStateException("no metadata found for index [" + indexRoutingTable.index() + "]");
+            }
+            IndexMetaData.Builder indexMetaDataBuilder = null;
+            for (IndexShardRoutingTable shardRoutings : indexRoutingTable) {
+                final ShardRouting primary = shardRoutings.primaryShard();
+                if (primary == null) {
+                    throw new IllegalStateException("missing primary shard for " + shardRoutings.shardId());
+                }
+                final int shardId = primary.shardId().id();
+                if (primary.primaryTerm() != indexMetaData.primaryTerm(shardId)) {
+                    assert primary.primaryTerm() > indexMetaData.primaryTerm(shardId) :
+                            "primary term should only increase. Index primary term ["
+                                    + indexMetaData.primaryTerm(shardId) + "] but primary routing is " + primary;
+                    if (indexMetaDataBuilder == null) {
+                        indexMetaDataBuilder = IndexMetaData.builder(indexMetaData);
+                    }
+                    indexMetaDataBuilder.primaryTerm(shardId, primary.primaryTerm());
+                }
+            }
+            if (indexMetaDataBuilder != null) {
+                if (metaDataBuilder == null) {
+                    metaDataBuilder = MetaData.builder(currentMetaData);
+                }
+                metaDataBuilder.put(indexMetaDataBuilder);
+            }
+        }
+        if (metaDataBuilder != null) {
```

Comment:
Can we increase increment the meta data version here? This comments says only master should do that (and did before): https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java#L437 Will the version not increase a second time once we get here https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java#L443 for example if we updated index meta data with some settings and this triggered this reroute?